File: src/main/scala/org/unipi/matrixgen/MatrixGenInputFormat.scala
package org.unipi.matrixgen

import java.io.IOException
import org.apache.hadoop.io.{NullWritable, Text}
import org.apache.hadoop.mapreduce.{Job, JobContext, InputSplit, RecordReader, TaskAttemptContext}
import org.apache.hadoop.mapreduce.lib.input.FileInputFormat

class MatrixGenInputFormat extends FileInputFormat[Text, NullWritable] {

  @throws[IOException]
  override def getSplits(job: JobContext): java.util.List[InputSplit] = {

    val numSplits: Int = job.getConfiguration.getInt(MatrixGenInputFormat.NUM_MAP_TASKS, -1)
    val splits: java.util.ArrayList[InputSplit] = new java.util.ArrayList[InputSplit]
    for (i <- 1 to numSplits) {
      splits.add(new FakeInputSplit())
    }

    splits
  }

  @throws[IOException]
  @throws[InterruptedException]
  override def createRecordReader(split: InputSplit, context: TaskAttemptContext): RecordReader[Text, NullWritable] = {
    val rr = new MatrixGenRecordReader
    rr.initialize(split, context)
    rr
  }
}

object MatrixGenInputFormat {

  val NUM_MAP_TASKS = "random.generator.map.tasks"
  val NUM_RECORDS_PER_TASK = "random.generator.num.records.per.map.task"

  def setNumMapTasks(job: Job, i: Int): Unit = {
    job.getConfiguration.setInt(NUM_MAP_TASKS, i)
  }

  def setNumRecordsPerTask(job: Job, i: Int): Unit = {
    job.getConfiguration.setInt(NUM_RECORDS_PER_TASK, i)
  }
}

File: src/main/java/org/unipi/matrixgen/MatrixGenInputFormat.java
package org.unipi.matrixgen;

import org.apache.hadoop.io.NullWritable;
import org.apache.hadoop.io.Text;

import org.apache.hadoop.mapreduce.Job;
import org.apache.hadoop.mapreduce.JobContext;
import org.apache.hadoop.mapreduce.InputSplit;
import org.apache.hadoop.mapreduce.RecordReader;
import org.apache.hadoop.mapreduce.TaskAttemptContext;
import org.apache.hadoop.mapreduce.lib.input.FileInputFormat;

import java.io.IOException;
import java.util.List;
import java.util.ArrayList;

class MatrixGenInputFormat extends FileInputFormat<Text, NullWritable> {

    private static final String NUM_MAP_TASKS = "random.generator.map.tasks";
    static final String NUM_RECORDS_PER_TASK = "random.generator.num.records.per.map.task";

    public List<InputSplit> getSplits(JobContext job) throws IOException {

        int numSplits = job.getConfiguration().getInt(NUM_MAP_TASKS, -1);

        ArrayList<InputSplit> splits = new ArrayList<>();

        for (int i = 0; i < numSplits; i++) {
            splits.add(new FakeInputSplit());
        }

        return splits;
    }

    public RecordReader<Text, NullWritable> createRecordReader(InputSplit split, TaskAttemptContext context)
            throws IOException, InterruptedException {
        MatrixGenRecordReader rr = new MatrixGenRecordReader();
        rr.initialize(split, context);
        return rr;
    }

    static void setNumMapTasks(Job job, int i) {
        job.getConfiguration().setInt(NUM_MAP_TASKS, i);
    }

    static void setNumRecordsPerTask(Job job, int i) {
        job.getConfiguration().setInt(NUM_RECORDS_PER_TASK, i);
    }
}